geode-core/src/test/java/org/apache/geode/distributed/internal/InternalLocatorTest.java

```diff
@@ -0,0 +1,157 @@
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.geode.distributed.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import org.junit.Test;
+
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.cache.internal.HttpService;
+import org.apache.geode.internal.cache.InternalCacheForClientAccess;
+import org.apache.geode.internal.cache.InternalRegionFactory;
+import org.apache.geode.internal.security.SecurableCommunicationChannel;
+import org.apache.geode.logging.internal.LoggingSession;
+import org.apache.geode.management.internal.BaseManagementService;
+
+public class InternalLocatorTest {
+  @Test
+  public void startClusterManagementServiceWithRestServiceEnabledInvokesStartManager()
+      throws IOException {
+    LoggingSession loggingSession = mock(LoggingSession.class);
+    DistributionConfigImpl distributionConfig = mock(DistributionConfigImpl.class);
+    when(distributionConfig.getJmxManager()).thenReturn(true);
+    when(distributionConfig.getLocators()).thenReturn("");
+    when(distributionConfig.getSecurableCommunicationChannels())
+        .thenReturn(new SecurableCommunicationChannel[] {});
+    when(distributionConfig.getSecurityAuthTokenEnabledComponents()).thenReturn(new String[] {});
+    InternalLocator internalLocator = new InternalLocator(0, loggingSession, null, null, null, null,
+        null, null, distributionConfig, null);
+    InternalCacheForClientAccess cache = mock(InternalCacheForClientAccess.class);
+    InternalRegionFactory regionFactory = mock(InternalRegionFactory.class);
+    when(cache.createInternalRegionFactory(RegionShortcut.REPLICATE)).thenReturn(regionFactory);
+    when(cache.getOptionalService(HttpService.class))
+        .thenReturn(Optional.of(mock(HttpService.class)));
+    when(cache.getCacheForProcessingClientRequests()).thenReturn(cache);
+    BaseManagementService managementService = mock(BaseManagementService.class);
+    BaseManagementService.setManagementService(cache, managementService);
+    internalLocator.setInternalCache(cache);
```

Comment:
Instead of adding a new test-only method, try this:
- Add a new `Supplier<InternalCache>` parameter to the `InternalLocator` constructor, and let the constructor store the supplier in a field for later use.
- Change `InternalLocator.getCache()` to call the supplier instead of calling `GemFireCacheImpl.getInstance()` directly.
- Change `InternalLocator.createLocator(…)` and `InternalLocatorIntegrationTest` to pass `GemFireCacheImpl::getInstance` to the `InternalLocator` constructor.
- In these new tests, pass the mock cache to the `InternalLocator` constructor.

Reply:
I was able to get rid of the new test-only method. But I did it in a different way that did not require the constructor to change. The method being tested now takes as a parameter the cache it should use. Let me know what you think.